cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -105,13 +105,115 @@ cartographer_ros_msgs::SubmapList MapBuilderBridge::GetSubmapList() {
       cartographer_ros_msgs::SubmapEntry submap_entry;
       submap_entry.submap_version = submaps->Get(submap_index)->num_range_data;
       submap_entry.pose = ToGeometryMsgPose(submap_transforms[submap_index]);
+      submap_entry.local_pose = ToGeometryMsgPose(submaps->Get(submap_index)->local_pose());
       trajectory_submap_list.submap.push_back(submap_entry);
     }
     submap_list.trajectory.push_back(trajectory_submap_list);
   }
   return submap_list;
 }
 
+visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+  visualization_msgs::MarkerArray trajectory_nodes_list;
+
+  for (int trajectory_id = 0;
+      trajectory_id < map_builder_.num_trajectory_builders();
+      ++trajectory_id) {
+    const auto trajectory_nodes = map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+
+    int i = 0;
+    for (const auto& node : trajectory_nodes){
+      visualization_msgs::Marker marker;
+      marker.id = i++;
+      marker.type = visualization_msgs::Marker::ARROW;
+      marker.header.stamp = ::ros::Time::now();
+      marker.header.frame_id = options_.map_frame;
+      marker.color.b = 1.0;
+      marker.color.a = 1.0;
+      marker.scale.x = 0.1;
+      marker.scale.y = 0.05;
+      marker.scale.z = 0.05;
+      marker.pose = ToGeometryMsgPose(node.pose);
+      trajectory_nodes_list.markers.push_back(marker);
+    }
+  }
+
+  return trajectory_nodes_list;
+}
+
+cartographer_ros_msgs::ConstraintVisualization MapBuilderBridge::GetConstraintsList() {
+  cartographer_ros_msgs::ConstraintVisualization constraint_visualization;
+
+  const std::vector<cartographer::transform::Rigid3d> submap_transforms =
+      map_builder_.sparse_pose_graph()->GetSubmapTransforms();
+
+  const auto trajectory_nodes = map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+
+  int i = 0;
+  for (const auto &constraint : constraints) {
+
+	visualization_msgs::Marker constraint_marker, residual_error_marker;
+
+    // creating constraint line strip marker (green)
+    constraint_marker.id = i++;
+    constraint_marker.type = visualization_msgs::Marker::LINE_STRIP;
+    constraint_marker.header.stamp = ::ros::Time::now();
+    constraint_marker.header.frame_id = options_.map_frame;   
+    constraint_marker.color.a = 1.0;
+    constraint_marker.scale.x = 0.02;
+
+    // creating residual error line strip marker (blue)
+    residual_error_marker.id = i++;
+    residual_error_marker.type = visualization_msgs::Marker::LINE_STRIP;
+    residual_error_marker.header.stamp = ::ros::Time::now();
+    residual_error_marker.header.frame_id = options_.map_frame;
+    residual_error_marker.color.a = 1.0;
+    residual_error_marker.scale.x = 0.02;
+
+    // creating points for the line strip markers
```

Comment:
Style: Use full sentences in comments, i.e. capitalize at the beginning and end with a period.